aggregator/src/aggregator/queue.rs

```diff
@@ -0,0 +1,712 @@
+    id_generator: TicketIdGenerator,
+}
+
+impl LIFORequestQueue {
+    /// Creates a new [`Self`].
+    ///
+    /// `concurrency` must be greater than zero.
+    pub fn new(concurrency: u32, depth: usize, meter: &Meter) -> Result<Self, Error> {
+        if concurrency < 1 {
+            return Err(Error::InvalidConfiguration(
+                "concurrency must be greater than 0",
+            ));
+        }
+
+        let (message_tx, message_rx) = mpsc::unbounded_channel();
+        let id_generator = Default::default();
+        let metrics = Metrics::new(meter);
+        let dispatcher = Some(Self::dispatcher(message_rx, concurrency, depth, metrics));
+
+        Ok(Self {
+            dispatcher,
+            id_generator,
+            dispatcher_tx: message_tx,
+        })
+    }
+
+    /// Spawns a task that dispatches permits to waiting [`Ticket`]s. Once a permit is dispatched,
+    /// the ticket holder may proceed.
+    ///
+    /// Tickets are dispatched in LIFO-order. An incoming ticket will either immediately receive
+    /// a ticket, be placed at the head of the queue, or will be rejected if there's no space in
+    /// the queue.
+    ///
+    /// Permits are returned to the dispatcher automatically when the [`OwnedSemaphorePermit`] is
+    /// dropped, allowing other Tickets to be dispatched.
+    fn dispatcher(
+        mut dispatcher_rx: mpsc::UnboundedReceiver<DispatcherMessage>,
+        concurrency: u32,
+        depth: usize,
+        metrics: Metrics,
+    ) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            // Use a BTreeMap to allow for cancellation (i.e. removal) of tickets in sublinear
+            // time.
+            let mut stack: BTreeMap<TicketId, PermitTx> = BTreeMap::new();
+
+            // Unwrap safety: conversion only fails on architectures where usize is less than
+            // 32-bits.
+            let semaphore = Arc::new(Semaphore::new(concurrency.try_into().unwrap()));
+
+            // Create a collection of permits which we'll dispatch to tickets.
+            // Unwrap safety: the semaphore is never closed.
+            let mut permits = Arc::clone(&semaphore)
+                .acquire_many_owned(concurrency)
+                .await
+                .unwrap();
+
+            loop {
+                let semaphore = Arc::clone(&semaphore);
+                select! {
+                    Some(message) = dispatcher_rx.recv() => {
+                        match message {
+                            DispatcherMessage::Enqueue(id, permit_tx) => {
+                                if let Some(permit) = permits.split(1) {
+                                    permit_tx.send(Ok(permit))
+                                } else if stack.len() < depth {
+                                    let result = stack.insert(id, permit_tx);
+                                    assert!(result.is_none(), "ticket IDs must be unique");
```

Comment:
FWIW it won't lead to a deadlock, since `.acquire()` gracefully handles the dispatcher channel being closed. That replica will just always 500. Which is still not really a desired outcome, so I'll change it to an `error!`.

Perhaps the health of the dispatcher task ought to be factored into the liveness check 🤔.

Reply:
> Perhaps the health of the dispatcher task ought to be factored into the liveness check 🤔.

I think this would be wise, but I'll consider it in a future PR.